horovod/common/ops/cuda_operations.cc

```diff
@@ -159,7 +159,10 @@ Status CUDAAllreduce::FinalizeCUDAQueue(const std::vector<TensorTableEntry>& ent
   // Claim a std::shared_ptr to the fusion buffer to prevent its memory from being reclaimed
   // during finalization.
   auto fusion_buffer = global_state_->fusion_buffer.GetBuffer(
-      first_entry.device, first_entry.context->framework(), global_state_->current_nccl_stream);
+      first_entry.device,
+      // JoinTensor doesn't have context or framework
+      first_entry.context ? first_entry.context->framework() : PYTORCH,
```

Comment:
Here PYTORCH isn't actually used afterwards, it can be any valid framework constant.
But I agree with your suggestion about delegating Tensor allocation to the framework, I'll try to do that.